### PR TITLE
8287835: Add support for additional float/double to integral conversion for x86

### PR DESCRIPTION
**Pull request by sviswa7**
Currently the C2 JIT only supports float -> int and double -> long conversion for x86. 
This PR adds the support for following conversions in the c2 JIT:
  float -> long, short, byte
  double -> int, short, byte

The performance gain is as follows.
Before the patch:
Benchmark                                       Mode  Cnt      Score       Error   Units
VectorFPtoIntCastOperations.microDouble2Byte   thrpt    3  32367.971 ±  6161.118  ops/ms
VectorFPtoIntCastOperations.microDouble2Int    thrpt    3  25825.251 ±  5417.104  ops/ms
VectorFPtoIntCastOperations.microDouble2Long   thrpt    3  59641.958 ± 17307.177  ops/ms
VectorFPtoIntCastOperations.microDouble2Short  thrpt    3  29641.505 ± 12023.015  ops/ms
VectorFPtoIntCastOperations.microFloat2Byte    thrpt    3  16271.224 ±  1523.083  ops/ms
VectorFPtoIntCastOperations.microFloat2Int     thrpt    3  59199.994 ± 14357.959  ops/ms
VectorFPtoIntCastOperations.microFloat2Long    thrpt    3  17169.197 ±  1738.273  ops/ms
VectorFPtoIntCastOperations.microFloat2Short   thrpt    3  14934.139 ±  2329.253  ops/ms

After the patch:
Benchmark                                       Mode  Cnt       Score       Error   Units
VectorFPtoIntCastOperations.microDouble2Byte   thrpt    3  115436.659 ± 21282.364  ops/ms
VectorFPtoIntCastOperations.microDouble2Int    thrpt    3   87194.395 ±  9443.106  ops/ms
VectorFPtoIntCastOperations.microDouble2Long   thrpt    3   59652.356 ±  7240.721  ops/ms
VectorFPtoIntCastOperations.microDouble2Short  thrpt    3  110570.719 ± 10401.620  ops/ms
VectorFPtoIntCastOperations.microFloat2Byte    thrpt    3  110028.539 ± 11113.137  ops/ms
VectorFPtoIntCastOperations.microFloat2Int     thrpt    3   59469.193 ± 18272.495  ops/ms
VectorFPtoIntCastOperations.microFloat2Long    thrpt    3   59897.101 ±  7249.268  ops/ms
VectorFPtoIntCastOperations.microFloat2Short   thrpt    3   86167.554 ±  8253.232  ops/ms

Please review.

Best Regards,
Sandhya

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287835](https://bugs.openjdk.org/browse/JDK-8287835): Add support for additional float/double to integral conversion for x86

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [996ee049](https://git.openjdk.java.net/jdk/pull/9032/files/996ee0494c80523c43bdf4c46a112699b0a229cc)
 * [Jatin Bhateja](https://openjdk.java.net/census#jbhateja) (@jatin-bhateja - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/9032/head:pull/9032` \
`$ git checkout pull/9032`

Update a local copy of the PR: \
`$ git checkout pull/9032` \
`$ git pull https://git.openjdk.java.net/jdk pull/9032/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9032`

View PR using the GUI difftool: \
`$ git pr show -t 9032`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/9032.diff">https://git.openjdk.java.net/jdk/pull/9032.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1889]: Again overlapping conditions. So T_LONG requires both: AVX512, avx512vl and avx512dq?

What about T_INT?

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:7298]: Do we need some of these conditions since you have them already in `match_rule_supported_vector()`?

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1877]: Overlapping conditions for the same types are confusing.

**Review by vnkozlov — COMMENTED** [review]
I assume it is support for "vector conversion".

Please, add IR framework test.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:1889]: T_INT doesn't need AVX512dq. Float to long conversion (T_LONG) uses evcvttps2qq, which needs AVX512dq.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:1877]: I will add comments and rephrase the checks to make it clearer.

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [reply src/hotspot/cpu/x86/x86.ad:1889]: Okay. I see that there are 2 instructions to support F2I by using avx or evex encoding. They cover all cases.
No you are introducing sub_integer and long types only for evex encoding.

You need comment that F2I is supported in all cases. For other integral types you need 512vl  and additionally 512dq for T_LONG.

Note, you don't need to check (UseAVX <= 2) because avx512vl bit is cleaned in such case. It is the same for VectorCastD2X code.

In such case I suggest:

```
  if (is_subword_type(bt) && !VM_Version::supports_avx512vl() ||
      (bt == T_LONG) && !VM_Version::supports_avx512vldq()) {
    return false;
  }
```

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1871]: Which asm instructions are required avx512vl? I don't see asserts in `assembler_x86.cpp`

Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad]: avx512vl support is needed only for vectors < 512 bit. I have corrected this in the predicate.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad]: Yes it was needed to select between the rules. On platforms that don't support avx512vl, we use AVX512 instructions only for 512-bit vectors and AVX instructions for < 64 byte vectors.

**Comment by sviswa7** [comment]
@vnkozlov I have implemented your review comments. The only item remaining is to add IR framework test.

**Review by vnkozlov — COMMENTED** [review]
Looks good. Will wait IR test before testing and approval.

**Comment by sviswa7** [comment]
@vnkozlov I have added the IR framework test case. Please take a look.

**Review by vnkozlov — APPROVED** [review]
Good. I will start testing.

**Review by vnkozlov — APPROVED** [review]
Results are good.
You need second review.

**Comment by sviswa7** [comment]
@vnkozlov Thanks a lot for the review and test.
@jatin-bhateja Could you please review this PR. It is an extension of your earlier work.

File: test/hotspot/jtreg/compiler/vectorapi/VectorFPtoIntCastTest.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorapi/VectorFPtoIntCastTest.java:45]: Unused declarations.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad]: D2X in general needs AVX 512 due to evcvttpd2qq.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:7349]: I will file a separate RFE for this.

**Comment by sviswa7** [comment]
@jatin-bhateja I have implemented your review comments. Please take a look.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad]: Yes, but can't we take D2F route to cast to integer and subword range. 
D2X -> D2F(AVX2) -> F2X.

**Comment by jatin-bhateja** [comment]
Thanks @sviswa7

**Review by jatin-bhateja — APPROVED** [review]

**Comment by sviswa7** [comment]
@vnkozlov Could I go ahead and integrate? There were some minor changes and code rearrangement after your last test. Please let me know.

**Comment by vnkozlov** [comment]
I submitted new testing.

**Comment by sviswa7** [comment]
Thanks a lot Vladimir!

**Comment by vnkozlov** [comment]
@sviswa7 testing results are good. You can push.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad]: Thanks @sviswa7 , for AVX we can use VCVTTPD2DQ to cast double precison lane to integer and subsequently to sub  words lanes. For casting to long we  do not have direct instruction.

File: test/hotspot/jtreg/compiler/vectorapi/VectorFPtoIntCastTest.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorapi/VectorFPtoIntCastTest.java:79]: Can you kindly all add special floating point cases NaN, +/-Inf, +/-0.0 to cover your special handling code changes.

Comment (sviswa7) [reply test/hotspot/jtreg/compiler/vectorapi/VectorFPtoIntCastTest.java:79]: @jatin-bhateja The test is only checking the IR node generation for x86. 
The rest of the actual functionality test is already covered under the following including the special cases:
compiler/codegen/TestByteDoubleVect.java compiler/codegen/TestByteFloatVect.java compiler/codegen/TestShortFloatVect.java compiler/codegen/TestShortDoubleVect.java compiler/codegen/TestLongFloatVect.java compiler/codegen/TestIntDoubleVect.java compiler/codegen/TestIntFloatVect.java
The general idea of this PR was to complement x86 FP to integral conversion along with https://git.openjdk.org/jdk/pull/7806 from Fei Gao.

**Pull request closed without merging**
